lib/compat/wordpress-6.5/navigation-block-hooks.php

```diff
@@ -0,0 +1,107 @@
+<?php
+add_filter( 'rest_prepare_wp_navigation', 'gutenberg_hook_blocks_into_core_navigation_rest_response', 10, 3 );
+add_action( 'rest_insert_wp_navigation', 'gutenberg_update_ignore_hooked_blocks_meta_core_navigation', 10, 3 );
+
+/**
+ * Updates the post meta with the list of ignored hooked blocks when the navigation is created or updated via the REST API.
+ *
+ * @param WP_Post $post Post object.
+ */
+function gutenberg_update_ignore_hooked_blocks_meta_core_navigation( $post ) {
+	if ( ! isset( $post->ID ) ) {
+		return;
+	}
+
+	$anchor_block_type    = 'core/navigation';
+	$first_child_position = 'first_child';
+	$last_child_position  = 'last_child';
+
+	$hooked_blocks                 = get_hooked_blocks();
+	$hooked_block_types_last_child = isset( $hooked_blocks[ $anchor_block_type ][ $last_child_position ] )
+		? $hooked_blocks[ $anchor_block_type ][ $last_child_position ]
+		: array();
+
+	$hooked_block_types_first_child = isset( $hooked_blocks[ $anchor_block_type ][ $first_child_position ] )
+		? $hooked_blocks[ $anchor_block_type ][ $first_child_position ]
+		: array();
+
+	$hooked_block_types_last_child  = apply_filters( 'hooked_block_types', $hooked_block_types_last_child, $last_child_position, $anchor_block_type, $post );
+	$hooked_block_types_first_child = apply_filters( 'hooked_block_types', $hooked_block_types_first_child, $first_child_position, $anchor_block_type, $post );
+
+	$ignored_hooked_blocks = array_merge( $hooked_block_types_last_child, $hooked_block_types_first_child );
```

Comment:
I finally managed to flesh out what I meant with my earlier suggestion to run the "entire" Block Hooks mechanism here, and extract the ignored hooked blocks from it. Here's what that would look like:

 ```suggestion
	// We run the Block Hooks mechanism so it will return the list of ignored hooked blocks
	// in the mock root Navigation block's metadata attribute.
	// We ignore the rest of the returned `$markup`; `$post->post_content` already has the hooked
	// blocks inserted, whereas `$markup` will have them inserted twice.
	$blocks         = parse_blocks( $post->post_content );
	$markup         = gutenberg_hook_first_last_children_into_core_navigation( $blocks, $post->ID );
	$root_nav_block = parse_blocks( $markup )[0];

	$ignored_hooked_blocks = isset( $root_nav_block['attrs']['metadata']['ignoredHookedBlocks'] )
		? $root_nav_block['attrs']['metadata']['ignoredHookedBlocks']
		: array();
```

I've tested this, and AFAICT, it works. LMK what you think about it. As stated elsewhere, I think the major benefit of this approach is that we won't need to manually sync it with the implementation of Block Hooks whenever it changes. OTOH, it adds a bit of overhead, as it runs the Block Hooks mechanism on a parsed block tree that already has them inserted (see inline comment); although this overhead is probably fairly small, as Navigation menu trees don't tend to have that many nodes.

(Personally, I probably lean slightly towards this approach (mostly based on some instances where blocks had been using low-level methods to re-implement existing higher-level ones; they would either miss some functionality or go  out of sync, see e.g. #52892 or #55811). But it's not a very strong preference at this point.)

Reply:
Great suggestion @ockham 🎉 . This keeps this implementation closer to the underlying mechanism of the Block Hooks API since we are using the helper functions within `gutenberg_hook_first_last_children_into_core_navigation`. As well this, it also keeps it consistent with every other time we interact with the core Navigations inner blocks. For these reasons I'm also leaning (heavily) towards update.

Having seen this, it seems to obvious I'm slightly annoyed I didn't think of it myself 😂 I think I've been staring at this code for too long now.

Reply:
Done in https://github.com/WordPress/gutenberg/pull/57754/commits/06f06490b1b33f1cdbc3e12260dc0cb25f9afc5b

I've done some testing and this tests well but more eyes on this the better imo.

Reply:
I think your approach was fine, and arguably even optimized for performance, so no shame in that! 😄